src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -14,35 +14,288 @@ namespace Microsoft.CodeAnalysis
+            // and anything else with a reverse dependency on them. If we don't already know our reverse transitive references, then we'll have to instead remove
+            // the cache entries instead of update them. We'll fetch this from the map, and use "null" to indicate the "we don't know and should remove the cache entry"
+            // instead
+            existingReverseTransitiveReferencesMap.TryGetValue(projectId, out var newReverseTranstiveReferences);
+
+            if (newReverseTranstiveReferences != null)
+            {
+                newReverseTranstiveReferences = newReverseTranstiveReferences.Add(projectId);
+            }
+
+            // We'll now loop through each entry in our existing cache and compute updates. We'll accumulate them into this builder.
+            var builder = existingReverseTransitiveReferencesMap.ToBuilder();
+
+            foreach (var projectIdToUpdate in existingReverseTransitiveReferencesMap.Keys)
+            {
+                existingReverseTransitiveReferencesMap.TryGetValue(projectIdToUpdate, out var existingReverseTransitiveReferences);
+
+                // The projects who need to have their caches updated are projectIdToUpdate (since we're obviously updating it!)
+                // and also anything that depended on us.
+                if (referencedProjectIds.Contains(projectIdToUpdate) || existingReverseTransitiveReferences?.Overlaps(referencedProjectIds) == true)
+                {
+                    // This needs an update. If we know what to include in, we'll union it with the existing ones. Otherwise, we don't know
+                    // and we'll remove any data from the cache.
+                    if (newReverseTranstiveReferences != null && existingReverseTransitiveReferences != null)
+                    {
+                        builder[projectIdToUpdate] = existingReverseTransitiveReferences.Union(newReverseTranstiveReferences);
+                    }
+                    else
+                    {
+                        // Either we don't know the full set of the new references being added, or don't know the existing set projectIdToUpdate.
+                        // In this case, just remove it
+                        builder.Remove(projectIdToUpdate);
+                    }
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        internal ProjectDependencyGraph WithProjectReferences(ProjectId projectId, IEnumerable<ProjectId> referencedProjectIds)
+        {
+            // This method we can't optimize very well: changing project references arbitrarily could invalidate pretty much anything. The only thing we can reuse is our
+            // actual map of project references for all the other projects, so we'll do that
+            return new ProjectDependencyGraph(_projectIds, _referencesMap.SetItem(projectId, referencedProjectIds.ToImmutableHashSet()));
         }
```

Comment:
tiny nit.  i would have liked this method above the honking big one.

Reply:
Intent was the AddAdditional and all helpers were together. If this looks like an afterthought...good. :smile: